tests/unit_tests/plugin_manager_test.py

```diff
@@ -0,0 +1,219 @@
+import re
+from unittest import mock
+
+import click
+
+from mmpy_bot import Message, Plugin, Settings, listen_to, listen_webhook
+from mmpy_bot.driver import Driver
+from mmpy_bot.plugins import PluginManager
+
+
+# Used in the plugin tests below
+class FakePlugin(Plugin):
+    """Hello FakePlugin.
+
+    This is a plugin level docstring
+    """
+
+    @listen_to("pattern", needs_mention=True)
+    def my_function(self, message, another_arg=True):
+        """This is the docstring of my_function."""
+        pass
+
+    @listen_to("direct_pattern", direct_only=True, allowed_users=["admin"])
+    def direct_function(self, message):
+        """Help direct function."""
+        pass
+
+    @listen_to("async_pattern")
+    @listen_to("another_async_pattern", direct_only=True)
+    async def my_async_function(self, message):
+        """Async function docstring."""
+        pass
+
+    @listen_to("click_command")
+    @click.command(help="Help string for the entire function.")
+    @click.option(
+        "--option", type=int, default=0, help="Help string for the optional argument."
+    )
+    def click_commmand(self, message, option):
+        """Extended docstring.
+
+        This docstring will have 'click --help' appended
+        """
+        pass
+
+    @listen_to("hi_custom", custom="Custom attribute")
+    def hi_custom(self, message):
+        """A custom function."""
+        pass
+
+    @listen_webhook("webhook_id")
+    def webhook_listener(self, event):
+        """A webhook function."""
+        pass
+
+
+def expand_func_names(f):
+    return [f] + f.siblings
+
+
+msg_listeners = {
+    re.compile("pattern"): expand_func_names(FakePlugin.my_function),
+    re.compile("direct_pattern"): expand_func_names(FakePlugin.direct_function),
+    re.compile("another_async_pattern"): expand_func_names(
+        FakePlugin.my_async_function
+    ),
+    re.compile("async_pattern"): expand_func_names(FakePlugin.my_async_function),
+    re.compile("hi_custom"): expand_func_names(FakePlugin.hi_custom),
+    # Click commands construct a regex pattern from the listen_to pattern
+    re.compile("^click_command (.*)?"): expand_func_names(FakePlugin.click_commmand),
+}
```

Comment:
Why are these redefined? Can't we just take the arguments passed to `listen_to`?

Reply:
Did you accidentally link the wrong file? I don't see the relevance of that line here. I think the relevant test was [this one](https://github.com/attzonko/mmpy_bot/blob/main/tests/unit_tests/event_handler_test.py#L59), and it did not require duplicating the listeners like this

Reply:
Sorry I read your comment in a different light. I was justifying the why regexp and click listeners are treated differently with a link to the code where click listeners are transformed into a 'catch-all' regexp.

As to why I re-defined the list of expected listeners... The original test was comparing information stored in `Plugin` instances with that on the event handler. Since there's now only one place where listeners are collected, the only way I could think to test this was by either repeating the logic already used by the `PluginManager` or hardcoding the expected result. The first option didn't make sense to me so I went with the second.

Reply:
I see. I still believe that the logic of finding the listeners on any given Plugin should be in a function on the Plugin itself, even if it will only be called by the PluginManager. That also removes the need for any code duplication here.

Reply:
That's doable but with it we need to do some extra work later when building the help page. i.e. we will need to traverse all the plugins every time, sort/group, etc... instead of just calling to the `PluginManager`.
I remember I started with this approach but ran into some situations either I duplicated information both in `Plugin` and `PluginManager` or I was constantly looping over all plugins.
If you think listener initialization should stay with `Plugin` we might need to bring the duplication back to avoid a possible performance hit with looping for almost every event.